### PR TITLE
Switch to Phusion as base image

### PR DESCRIPTION
**Pull request by fbartels**
Relates to https://github.com/Kopano-dev/kopano-one-container/issues/8

### CODE REVIEW

File: docs/docker-compose/docker-compose.yml [hunk 1]
```diff
@@ -3,13 +3,15 @@ version: "3.5"
 services:
   kopano-one:
     image: ${docker_repo:-kopano}/kopano-one:${ONE_VERSION:-latest}
-    read_only: true
+    #read_only: true
     restart: unless-stopped
     volumes:
+      - ./config:/etc/kopano/
```
[the remark this answers is not in the document]
Comment (longsleep) [reply docs/docker-compose/docker-compose.yml:12]: I don't understand how this is possible.

A volume mount to /etc/kopano should replace the symlink at the same place and leave the target of the sym link untouched. Thats not the case?

File: Dockerfile [hunk 2]
```diff
@@ -14,29 +14,55 @@ LABEL maintainer=development@kopano.io \
...
+	find /etc/mysql/ -name '*.cnf' -print0 \
+		| xargs -0 grep -lZE '^(bind-address|log|user\s)' \
+		| xargs -rt -0 sed -Ei 's/^(bind-address|log|user\s)/#&/'&& \
+# don't reverse lookup hostnames, they are usually another container
+	echo '[mysqld]\nskip-host-cache\nskip-name-resolve' > /etc/mysql/conf.d/docker.cnf
 
-ARG VCS_REF
-LABEL org.label-schema.vcs-ref=$VCS_REF
+# retrieve mariadb entrypoint
+RUN curl -o /usr/local/bin/mariadb-entrypoint.sh https://raw.githubusercontent.com/MariaDB/mariadb-docker/master/docker-entrypoint.sh && \
```
Comment (longsleep) [inline Dockerfile:45]: This link should point to a specific Git revision or to a specific tag, to  avoid surprises.

File: Makefile [hunk 3]
```diff
@@ -6,27 +6,38 @@ SHELL := bash
 MAKEFLAGS += --warn-undefined-variables
 MAKEFLAGS += --no-builtin-rules
 
-docker_repo := kopano
+DOCKER_REPO := kopano
 
-vcs_ref := $(shell git rev-parse --short HEAD)
-date := $(shell date '+%Y%m%d%H%M%S')
+VCS_REF := $(shell git rev-parse --short HEAD)
+DATE := $(shell date '+%Y%m%d%H%M%S')
```
Comment (longsleep) [inline Makefile:12]: I suggest to use a different variable name here (e.g. TAG).

File: README.md [hunk 4]
```diff
@@ -1,21 +1,39 @@
 # Kopano ONE Container
 
+!! This project is still a work in progress !!
```
Comment (longsleep) [inline README.md:3]: Seems not really required - i am not a big fan of these markers.

File: docs/docker-compose/config/kidm/ldif/main.d/10-main.ldif [hunk 5]
```diff
@@ -0,0 +1,4331 @@
...
+objectClass: device
+objectClass: ipHost
+objectClass: kopano-server
+objectClass: top
+kopanoContainsPublic: 0
+kopanoFilePath: /var/run/kopano/server.sock
+kopanoHttpPort: 236
+kopanoSslPort: 237
+
+# user2, users, {{.MailDomain}}
```
Comment (longsleep) [inline docs/docker-compose/config/kidm/ldif/main.d/10-main.ldif:81]: Sould this really have users in by default?

File: docs/docker-compose/config/konnectd.cfg [hunk 6]
```diff
@@ -0,0 +1,2 @@
+oidc_issuer_identifier=https://kopano.demo:9443
+insecure = yes
```
Comment (longsleep) [inline docs/docker-compose/config/konnectd.cfg:2]: Uhm - why is this needed?

File: docs/docker-compose/config/konnectd.cfg [hunk 7]
```diff
@@ -0,0 +1,2 @@
+oidc_issuer_identifier=https://kopano.demo:9443
```
Comment (longsleep) [inline docs/docker-compose/config/konnectd.cfg:1]: Port seems not useful. Why does it not point to 443 ?

File: docs/docker-compose/config/server.cfg [hunk 8]
```diff
@@ -0,0 +1,11 @@
+enable_sso = yes
+# (external) address of this Kopano system
+kcoidc_issuer_identifier = https://kopano.demo:9443
```
Comment (longsleep) [inline docs/docker-compose/config/server.cfg:3]: I don't understand why u use port 9443 everywhere (only have it listen on port 9443, but expose it on 443 to the exsternal world together with the external hostname).

File: docs/docker-compose/config/server.cfg [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+# This option should not be set to "yes" in production environments!
+kcoidc_insecure_skip_verify=yes
```
Comment (longsleep) [inline docs/docker-compose/config/server.cfg:5]: Please let's not shop insecure defaults.

File: docs/docker-compose/docker-compose.yml [hunk 10]
```diff
@@ -3,21 +3,27 @@ version: "3.5"
 services:
   kopano-one:
     image: ${docker_repo:-kopano}/kopano-one:${ONE_VERSION:-latest}
-    read_only: true
+    #read_only: true
+    ports:
+      - "9080:9080"
+      - "9443:9443"
```
Comment (longsleep) [inline docs/docker-compose/docker-compose.yml:9]: The exposed ports should be 80 and 443 imo.

File: runit/kopano-kidmd/run [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+ensure_config kidmd.cfg
+ensure_dir /etc/kopano/kidm/ldif/main.d/
+
+# read settings from kidmd.cfg into env (emulates systemds EnvironmentFile)
+sed s/\ *=\ */=/g "/etc/kopano/kidmd.cfg" > /tmp/kidmd-env
+# shellcheck disable=SC2046
+export $(grep -v '^#' /tmp/kidmd-env | xargs -d '\n')
+
+exec /sbin/setuser kopano /usr/sbin/kopano-kidmd serve
```
Comment (longsleep) [inline runit/kopano-kidmd/run:14]: Should not run as user `kopano`. Use nobody or similar please.

File: runit/kopano-kustomerd/run [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+ensure_dir /etc/kopano/licenses
+ensure_dir /run/kopano-kustomerd
+chown kopano /run/kopano-kustomerd
+
+# read settings from kustomerd.cfg into env (emulates systemds EnvironmentFile)
+sed s/\ *=\ */=/g "/etc/kopano/kustomerd.cfg" > /tmp/kustomer-env
+# shellcheck disable=SC2046
+export $(grep -v '^#' /tmp/kustomer-env | xargs -d '\n')
+
+exec /sbin/setuser kopano /usr/sbin/kopano-kustomerd serve
```
Comment (longsleep) [inline runit/kopano-kustomerd/run:16]: Should not run as user kopano.

File: runit/kopano-server/run [hunk 13]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+set -euo pipefail
+
+source ../kopano-helper.sh
+
+ensure_config server.cfg
+ensure_config ldap.cfg
+
+exec 2>&1
+exec /usr/sbin/kopano-server -F
```
Comment (longsleep) [inline runit/kopano-server/run:10]: Do we need to start it as root? Imo it is better to start is as user kopano directly.

File: runit/kopano-smtpstd/run [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+ensure_config smtpstd.cfg
+ensure_dir /var/lib/kopano-smtpstd
+chown kopano:kopano /var/lib/kopano-smtpstd
+
+# read settings from smtpstd.cfg into env (emulates systemds EnvironmentFile)
+sed s/\ *=\ */=/g "/etc/kopano/smtpstd.cfg" > /tmp/smtpstd-env
+# shellcheck disable=SC2046
+export $(grep -v '^#' /tmp/smtpstd-env | xargs -d '\n')
+
+exec /sbin/setuser kopano /usr/sbin/kopano-smtpstd serve
```
Comment (longsleep) [inline runit/kopano-smtpstd/run:15]: Please no user kopano. Use nobody or a specific system user for all services except the ones which actually need direct access to the kopano-server socket.

File: runit/php/run [hunk 15]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+mkdir /run/php/
+exec /usr/sbin/php-fpm7.4 --nodaemonize -R
```
Comment (longsleep) [inline runit/php/run:3]: Not sure, but i assume this switches to some different user as configured (the workers).

**Review by longsleep — CHANGES_REQUESTED** [review]
I think this is going to be a pretty nice container :+1:

File: docs/docker-compose/docker-compose.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (fbartels) [reply docs/docker-compose/docker-compose.yml:12]: No, definitely not the behaviour that I expected as well, but like I wrote mounting over a symlinked dir resulted in the mounted dir also "overwriting" the symlink destination

File: README.md (hunk 4 above)
Comment (fbartels) [reply README.md:3]: Yes, I get your point. I wanted to have a marked to say "this is not yet finished", but on the other hand this could keep people back from providing feedback now that the repo is public.

File: docs/docker-compose/config/konnectd.cfg (hunk 6 above)
Comment (fbartels) [reply docs/docker-compose/config/konnectd.cfg:2]: This is just a file for demonstration in the documentation

File: runit/php/run (hunk 15 above)
Comment (fbartels) [reply runit/php/run:3]: yes, in the end its all running with www-data:

```
www-data      92  0.0  0.0 228960  7740 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data      95  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data      96  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data      97  0.0  0.0 228960  7844 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data      98  0.0  0.0 228960  7844 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data     100  0.0  0.0 228960  7844 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data     103  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data     104  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data     110  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data     115  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-kdav
www-data     116  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     117  0.0  0.0 228960  7840 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     118  0.0  0.0 228960  7844 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     119  0.0  0.0 228960  7844 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     120  0.0  0.0 228960  7844 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     121  0.0  0.0 228960  7844 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     123  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     124  0.0  0.0 228960  7848 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     125  0.0  0.0 228960  7852 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     127  0.0  0.0 228960  7852 ?        S    09:27   0:00 php-fpm: pool kopano-webapp
www-data     128  0.0  0.0 229076  7852 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     129  0.0  0.0 229076  7852 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     130  0.0  0.0 229076  7852 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     132  0.0  0.0 229076  7728 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     133  0.0  0.0 229076  7852 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     134  0.0  0.0 229076  7856 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     135  0.0  0.0 229076  7856 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     136  0.0  0.0 229076  7852 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     137  0.0  0.0 229076  7852 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
www-data     139  0.0  0.0 229076  7852 ?        S    09:27   0:00 php-fpm: pool kopano-z-push
```

File: runit/kopano-server/run (hunk 13 above)
Comment (fbartels) [reply runit/kopano-server/run:10]: Unfortunately at the moment it needs to stay like this, but there is also little harm since kopano-server reexecs itself and then runs as the user kopano.

When directly running the binary as the user kopano the following is logged:

```
root@76c3f3418c84:/etc/kopano# /sbin/setuser kopano /usr/sbin/kopano-server -F
2021-06-09T09:39:22.804711: [info   ] Audit logging not enabled.
2021-06-09T09:39:22.804748: [=======] Starting kopano-server version 9.0.13 (pid 833 uid 999)
2021-06-09T09:39:22.804774: [info   ] Using epoll events
2021-06-09T09:39:23.189252: [warning] K-1558: Unable to limit socket 0.0.0.0%lo:236 to lo: Operation not permitted
2021-06-09T09:39:23.189270: [info   ] Listening on 0.0.0.0%lo:236 (fd 7)
2021-06-09T09:39:23.575050: [warning] K-1558: Unable to limit socket [::]%lo:236 to lo: Operation not permitted
2021-06-09T09:39:23.575064: [info   ] Listening on [::]%lo:236 (fd 8)
2021-06-09T09:39:23.961872: [debug  ] K-1562: changed owner "/var/run/kopano/prio.sock" to kopano:kopano mode 660
2021-06-09T09:39:23.961885: [info   ] Listening on unix:/var/run/kopano/prio.sock (fd 9)
2021-06-09T09:39:24.347589: [debug  ] K-1562: changed owner "/var/run/kopano/server.sock" to kopano:kopano mode 666
2021-06-09T09:39:24.347601: [info   ] Listening on unix:/var/run/kopano/server.sock (fd 10)
2021-06-09T09:39:24.349499: [debug  ] Preparing relaunch with allocator lib libtcmalloc_minimal.so.4
2021-06-09T09:39:24.352692: [debug  ] Reexecing /usr/sbin/kopano-server
2021-06-09T09:39:24.365252: [info   ] Audit logging not enabled.
2021-06-09T09:39:24.365289: [=======] Starting kopano-server version 9.0.13 (pid 833 uid 999)
2021-06-09T09:39:24.365308: [info   ] Using epoll events
2021-06-09T09:39:24.368483: [warning] K-1558: Unable to limit socket 0.0.0.0%lo:236 to lo: Operation not permitted
2021-06-09T09:39:24.368496: [error  ] K-1559: bind 0.0.0.0%lo:236: Address already in use
2021-06-09T09:39:24.371630: [warning] K-1558: Unable to limit socket [::]%lo:236 to lo: Operation not permitted
2021-06-09T09:39:24.371638: [error  ] K-1559: bind [::]%lo:236: Address already in use

An error occurred: no access (0xffffff9e). Please check logfile file:- for details.
```

File: docs/docker-compose/config/server.cfg (hunk 8 above)
Comment (fbartels) [reply docs/docker-compose/config/server.cfg:3]: That is mostly because I have already something running on port 443 on my local system.
